packages/time-oracle/src/alarms.rs

```diff
@@ -11,151 +13,190 @@ fn as_seconds(from: Timestamp) -> TimeSeconds {
+        alarms_selection(self.storage, &self.alarms, as_seconds(ctime))
+    }
 }
 
-impl<'a> Alarms<'a> {
-    pub fn new(namespace_alarms: &'a str, namespace_index: &'a str) -> Self {
-        let indexes = AlarmIndexes {
-            alarms: MultiIndex::new(|_, d| *d, namespace_alarms, namespace_index),
-        };
+pub struct AlarmsMut<'storage> {
+    storage: &'storage mut dyn Storage,
+    alarms: IndexedMap,
+    in_delivery: Deque<'static, Addr>,
+}
 
-        let alarms = IndexedMap::new(namespace_alarms, indexes);
+impl<'storage> AlarmsMut<'storage> {
+    pub fn new(
+        storage: &'storage mut dyn Storage,
+        namespace_alarms: &'static str,
+        namespace_index: &'static str,
+        namespace_in_delivery: &'static str,
+    ) -> Self {
+        Self {
+            storage,
+            alarms: indexed_map(namespace_alarms, namespace_index),
+            in_delivery: Deque::new(namespace_in_delivery),
+        }
+    }
 
-        Self { alarms }
+    pub fn add(&mut self, subscriber: Addr, time: Timestamp) -> Result<(), AlarmError> {
+        self.add_internal(subscriber, as_seconds(time))
     }
 
-    pub fn add(
-        &self,
-        storage: &mut dyn Storage,
-        addr: Addr,
-        time: Timestamp,
-    ) -> Result<(), AlarmError> {
-        self.alarms.save(storage, addr, &as_seconds(time))?;
-        Ok(())
+    pub fn ensure_no_in_delivery(&mut self) -> Result<&mut Self, AlarmError> {
+        self.in_delivery
+            .is_empty(self.storage)?
+            .then_some(self)
+            .ok_or(AlarmError::NonEmptyAlarmQueue)
     }
 
-    pub fn remove(&self, storage: &mut dyn Storage, addr: Addr) -> Result<(), AlarmError> {
-        self.alarms.remove(storage, addr)?;
-        Ok(())
+    pub fn out_for_delivery(&mut self, subscriber: Addr) -> Result<(), AlarmError> {
+        self.alarms.remove(self.storage, subscriber.clone())?;
+
+        self.in_delivery
+            .push_back(self.storage, &subscriber)
+            .map_err(Into::into)
     }
 
-    pub fn alarms_selection<'b>(
-        &self,
-        storage: &'b dyn Storage,
-        ctime: Timestamp,
-    ) -> impl Iterator<Item = Result<(Addr, TimeSeconds), AlarmError>> + 'b
-    where
-        'a: 'b,
-    {
+    pub fn last_delivered(&mut self) -> Result<(), AlarmError> {
+        self.in_delivery
+            .pop_front(self.storage)
+            .map(|maybe_alarm: Option<Addr>| debug_assert!(maybe_alarm.is_some()))
+            .map_err(Into::into)
+    }
+
+    pub fn last_failed(&mut self, now: Timestamp) -> Result<(), AlarmError> {
+        self.in_delivery.pop_front(self.storage).map_err(Into::into).and_then(|maybe_alarm: Option<Addr>| {
+            maybe_alarm.ok_or(AlarmError::ReplyOnEmptyAlarmQueue)
+        }).and_then(|subscriber: Addr| self.add_internal(subscriber, as_seconds(now) - /* Minus one second, to ensure it can be run within the same block */ 1))
```

Comment:
let's unify the assertions in both cases - now one results in an error, the other - in a debug statement. Since we deal with an external component (wasmd) that may/may not obey the contract then I suggest failing with errors in both cases. Similarly to the way we do before start delivering a batch with alarms.